### ISSUE TITLE
Ineligible users could partake in a contribution

### ISSUE THREAD

**Issue by c4-submissions** [code-423n4/2023-10-party-findings#292]
# Lines of code

https://github.com/code-423n4/2023-10-party/blob/b23c65d62a20921c709582b0b76b387f2bb9ebb5/contracts/crowdfund/InitialETHCrowdfund.sol#L275

# Vulnerability details

## Impact

Users could sidestep the attempt of gatekeepers disallowing them from contributing to a crowdfund.

> NB: Most protocols that implement either a blocking or `allow/disallow listing mechanism` do this for legal actions, since with the whole legislation onboarding the web3 space, governmental bodies could have flagged an address as malicious and as such require that the address should not be allowed to integrate with `PartyDao's `systems, but currently there is no sure way of doing this as one can side step this via `contributeFor()` _which is a legal case in the making for the DAO_.

## Proof of Concept

First, take a look at [IGateKeeper.sol#L3-L16](https://github.com/code-423n4/2023-10-party/blob/b23c65d62a20921c709582b0b76b387f2bb9ebb5/contracts/gatekeepers/IGateKeeper.sol#L5-L15)

```solidity
// Interface for a gatekeeper contract used for private crowdfund instances.
interface IGateKeeper {
  /// @notice Check if a participant is eligible to participate in a crowdfund.
  /// @param participant The address of the participant.
  /// @param id The ID of the gate to eligibility against.
  /// @param userData The data used to check eligibility.
  /// @return `true` if the participant is allowed to participate, `false` otherwise.
  function isAllowed(
    address participant,
    bytes12 id,
    bytes memory userData
  ) external view returns (bool);
}
```

So one can see that the address passed should be the participant and a check should be made if they are eligible/allowed to participate in the crowd fund.

Now while contributing, be it via `contribute()` or `contributFor()` the internal[ `_contribute()`](https://github.com/code-423n4/2023-10-party/blob/b23c65d62a20921c709582b0b76b387f2bb9ebb5/contracts/crowdfund/InitialETHCrowdfund.sol#L275) function eventually gets called

```solidity
function _contribute(
  address payable contributor,
  address delegate,
  uint96 amount,
  uint256 tokenId,
  bytes memory gateData
) private returns (uint96 votingPower) {
  // Require a non-null delegate.
  if (delegate == address(0)) {
    revert InvalidDelegateError();
  }

  // Must not be blocked by gatekeeper.
  IGateKeeper _gateKeeper = gateKeeper;
  if (_gateKeeper != IGateKeeper(address(0))) {
    //@audit-issue
    if (!_gateKeeper.isAllowed(msg.sender, gateKeeperId, gateData)) {
      revert NotAllowedByGateKeeperError(
        msg.sender,
        _gateKeeper,
        gateKeeperId,
        gateData
      );
    }
  }

  votingPower = _processContribution(contributor, delegate, amount);

  // OK to contribute with zero just to update delegate.
  if (amount == 0) return 0;

  if (tokenId == 0) {
    // Mint contributor a new party card.
    party.mint(contributor, votingPower, delegate);
  } else if (disableContributingForExistingCard) {
    revert ContributingForExistingCardDisabledError();
  } else if (party.ownerOf(tokenId) == contributor) {
    // Increase voting power of contributor's existing party card.
    party.increaseVotingPower(tokenId, votingPower);
  } else {
    revert NotOwnerError(tokenId);
  }
}
```

As seen, If the gatekeeper is present it only checks the `msg.sender` and not the contributor to be in the allowed list, issue here is that context seems to have been wrongly put, since the hardcoded `msg.sender` is only right for `contribute()` and not `contributeFor()`, cause while both functions are to participate in a contribution the former is for oneself while the latter could be done for a different user.

With the above in mind, we can see that even if a user is not allowed to participate, they can just have someone who's allowed to to executee `contributor()` and pass them as the _contributor_ essentially allowing them to sidestep the allow/disallow mechanism that's to be followed.

as such one can delegate another person so as to side step this issue with gatekeepers not allowing them. Question: If the gatekeeper is present, while contributing there is a check that the msg.sender is allowed, I believe issue is that since the check is only applied to the msg.sender then someone that's allowed can just call this function and specify a person that's not allowed as the contributor, effectively sidestepping this, no? Title: People that are not allowed by the gatekeeper could still contribute, recommendation would be that while calling this function via `contributeFor()` a check should be made for the `contributor` instead and not `msg.sender`

Additionally if the functionality of this is not to help legal compliance purposes and only to create a private crowdfunding group then the code implementation _**still goes against**_ what's been clarified in [the docs](https://partydao.notion.site/PartyBid-Knowledge-Base-017342aca2be4b5bb362b05c71d362b0), cause even non-specified addresses could join the party... quoting docs below.

> - **Can I make a private Party?**

    Parties can be public or private. In public Parties, anyone can contribute. In private Parties, only specific people can contribute. There are various ways to make a Party private:

- **Allowlist**: Only specified addresses can join the Party.

## Tool used

Manual Review

## Recommended Mitigation Steps

Context in cases like this really matters, while contributing the `contributor` needs to be checked to be in the `allowList` instead of `msg.sender` since that is the final personnel participating, this way while a user tries contributing for himself, he already is the _contributor_ , now whenever `contributFor()` is called for another user there should always be an assurance that the user is in the `allowList` before their contribution gets validated.

TLDR: Do not hardcode the allowlisting check to `msg.sender`, it should be dependent on the context, i.e if `contributeFor()` is being executed then the contributor should be checked to be in the allowed list instead.

## Assessed type

Context
